astropy/io/fits/hdu/compressed/compressed.py

```diff
@@ -364,6 +366,28 @@ def __init__(
         self._orig_bscale = self._bscale
         self._orig_bitpix = self._bitpix
 
+        if (
+            self._bitpix > 0
+            and "BLANK" not in self._header
+            and "ZBLANK" not in self._header
+        ):
+            # check for column named "ZBLANK"
+            for i in range(1, self._header["TFIELDS"] + 1):
+                if self._header[f"TTYPE{i}"] == "ZBLANK":
```

Comment:
> Or is the header value here the one that the masked data need to be set to after decompression (which _should_ be `BLANK` according to the standard, but is permitted to be taken from `ZBLANK` instead)? 

That is correct.  `ZBLANK` can be different for each tile, in which case there is no need for a `ZBLANK` value in the header.  But after decompression, if there are blank values in the integer image, there must be a `BLANK` keyword in the uncompressed header that specifies what value should be used for blank pixels.  After uncompression, that blank value has to be the same for the whole image (since there is no memory of which pixels were in different tiles).

I think this an error in the FITS standard.  It should be required to have a `BLANK` value specified for integer images if `ZBLANK` is used.  Otherwise the reader must either invent a value for `BLANK` or must just ignore blanks.  (Getting the FITS standard changed is much harder than changing `astropy` though!)

> In that case this check seems correct, except that the standard does not require `[Z]BLANK` to be present if there actually are no undefined pixel in the image. Maybe having a `ZBLANK` column then is confusing, but it still seems like an allowed usage. But to confirm this, one would need to check if `blank_mask` in `decompress_image_data_section` is empty or not...

In principle this is true.  However, we must be able to generate the header for the uncompressed image without actually reading all the compressed data and scanning it for blanks.  That would be a huge performance hit (e.g., for the `section` functionality, which reads only a part of the compressed data).  So the only knowledge we have is whether `ZBLANK` was specified or not, and the only possible response if we find `ZBLANK` is to include a value for `BLANK`.

Reply:
Yes, I expected this would not be feasible at this stage, but also thought that if there are no undefined pixels no data would actually be changed anyway. The only catch is that when picking the default blank value here, some values _might_ match it by accident, but since this can be considered a broken, or at least poorly setup file, this should be acceptable. The warning is there to alert the user to the problem.